src/AI/reinforcementAI.js

```diff
@@ -0,0 +1,258 @@
+    // This will return array of SA indicies
+    return as;
+  }
+  let allowedSAIndicies = allowedSAIndicies(S); // []
+  maxSAIndex = argMax(QValueTable(allowedSAIndicies))
+
+Step 3: Apply action, let the game engine run until your next turn.
+  actionObj = SAIndexToActionObject(maxSAIndex)
+  store.dispatch({type: 'PLAY_CARD', cardToPlay: actionObj})
+
+Step 4: Update Q Value Table (agent learning)
+  Q(A, S) = Q(A, S) + alpha * [R(S) + max_A(Q(A, S')) - Q(A, S)]
+  let max_A = -inf;
+  let allowedSAIndicies = allowedSAIndicies(S'); // []
+  max_A = max(QValueTable[allowedSAIndicies])
+  QValueTable[maxSAIndex] = QValueTable[maxSAIndex] + alpha * [R(== -100 when dead) + max_A - QValueTable[maxSAIndex]]
+
+Step 5: Repeat Step 2 until game ends/training complete
+*/
+import math from 'mathjs';
+
+export default class ReinforcementAI {
+  constructor(stateSizeArr, actionSizeArr) {
+    this.QValueTable = [];
+    this.stateSizeArr = stateSizeArr;
+    this.actionSizeArr = actionSizeArr;
+    this.SASizeArr = [];
+    this.SASizeArr = this.SASizeArr.concat(stateSizeArr);
+    this.SASizeArr = this.SASizeArr.concat(actionSizeArr);
+    this.lastSAVector = -1;
+    this.alpha = 0.5;
+  }
+
+  initialize() {
+    this.QValueTable = math.random(this.SASizeArr);
+  }
+
+  getBestAction(reduxState) {
+    let SObject = this.reduxStateToSObject(reduxState);
+    let allowedSAVectors = this.allowedSAVectors(SObject);
+    let bestSAVector = this.getMaxQValueSAVectorGivenSAVectors(allowedSAVectors);
+    this.lastSAVector = bestSAVector;
+    return this.SAVectorToActionObject(bestSAVector);
+  }
+
+  learn(nextReduxState) {
+    let nextSObject = this.reduxStateToSObject(nextReduxState);
+    let allowedSAIndicies = this.allowedSAVectors(nextSObject);
+    let max_A = this.getMaxQValueGivenSAVectors(allowedSAIndicies);
+    this.QValueTable[this.lastSAVector] = this.QValueTable[this.lastSAVector] +
+      this.alpha * [this.reward(nextSObject) + max_A - this.QValueTable[this.lastSAVector]]
+  }
+
+  // Private functions
+  /*
+  {
+    player0dead: reduxState.players[0].dead,
+    player0lastCardId: player0lastCardId,
+    player0holdingCard0: reduxState.players[0].holdingCards[0],
+    player0holdingCard1: reduxState.players[0].holdingCards[1],
+  }
+  */
+  allowedSAVectors(SObject) {
+    // Represent index as 1d vector, math.subset([[0, 1], [2, 3]], math.index(1, 0)) // 2
+    let SVector = this.SObjectToSVector(SObject);
+    let AVectors = [], SAIndicies = [];
+    if (SObject.player0dead) {
+      return as;
+    }
+    var playAgainst = [], guess = [];
+    // SObject.player0card0
+    if (SObject.player0holdingCard0 === 1) {
+      playAgainst = [2, 3, 4];
+      guess = [2, 3, 4, 5, 6, 7, 8];
+    } else if (SObject.player0holdingCard0 === 8 || SObject.player0holdingCard0 === 7 || SObject.player0holdingCard0 === 4) {
+      playAgainst = [1];
+      guess = [2, 3, 4, 5, 6, 7, 8];
+    } else {
+      playAgainst = [2, 3, 4];
```

Comment:
Should not hard code possible moves by assuming RF AI player as player one. Didn't I change this already?

  allowedSAVectors(SObject, currentPlayerId, playerStatus) {
    // Represent index as 1d vector, math.subset([[0, 1], [2, 3]], math.index(1, 0)) // 2
    let SVector = this.SObjectToSVector(SObject);
    let AVectors = [], SAIndicies = [];
    if (SObject.player0dead) {
      return as;
    }
    var playAgainst = [1, 2, 3, 4], guess = [2, 3, 4, 5, 6, 7, 8];
    // Remove players who are dead or protected
    for (var i = playerStatus.length-1; i > -1; i--){
      if (playerStatus[i].dead || playerStatus[i].protected){
    		playAgainst.splice(i,1);    		
      }    	
    }
    var playerID_ind = playAgainst.indexOf(currentPlayerId);
    var playAgainst_copy = playAgainst.slice();
    // SObject.player0card0
    if (SObject.player0holdingCard0 === 1) {
      playAgainst.splice(playerID_ind, 1); // remove self player ID
    } else if (SObject.player0holdingCard0 === 8 || SObject.player0holdingCard0 === 7 || SObject.player0holdingCard0 === 4) {
      playAgainst = [currentPlayerId];
    } else {
      playAgainst.splice(playerID_ind, 1); // remove self player ID
    }
    AVectors = AVectors.concat(this.generateActionVectors(SObject.player0holdingCard0, playAgainst, guess));
    let actionObjects1 = this.AIndiciesToActionObjects(AVectors);

    // SObject.player0card1
    playAgainst = playAgainst_copy;
    if (SObject.player0holdingCard1 === 1) {
      playAgainst.splice(playerID_ind, 1); // remove self player ID
    } else if (SObject.player0holdingCard1 === 8 || SObject.player0holdingCard1 === 7 || SObject.player0holdingCard1 === 4) {
      playAgainst = [currentPlayerId];
    } else {
      playAgainst.splice(playerID_ind, 1); // remove self player ID
    }
    AVectors = AVectors.concat(this.generateActionVectors(SObject.player0holdingCard1, playAgainst, guess));
    let actionObjects2 = this.AIndiciesToActionObjects(AVectors);
    return this.combineSAVectors(SVector, AVectors);
  }